packages/lesswrong/components/posts/BookmarksList.tsx

```diff
@@ -1,29 +1,48 @@
-import { registerComponent, Components } from '../../lib/vulcan-lib';
 import React from 'react';
+import { registerComponent, Components } from '../../lib/vulcan-lib';
 import { useCurrentUser } from '../common/withUser';
+import { useMulti } from '../../lib/crud/withMulti';
 import withErrorBoundary from '../common/withErrorBoundary';
+import sortBy from 'lodash/sortBy';
+import findIndex from 'lodash/findIndex';
 
-const BookmarksList = ({limit=null}: {
-  limit?: number|null
+const BookmarksList = ({limit=20, hideLoadMore=false}: {
+  limit?: number,
+  hideLoadMore?: boolean,
 }) => {
   const currentUser = useCurrentUser();
-  const { PostsItem2 } = Components
-
-  let bookmarkedPosts = currentUser?.bookmarkedPosts || []
-  let truncated = false;
-  bookmarkedPosts = [...bookmarkedPosts].reverse();
-  if (limit) {
-    bookmarkedPosts = bookmarkedPosts.slice(0, limit)
-    truncated = true;
-  }
+  const { PostsItem2, LoadMore, Loading } = Components
+  
+  const {results: bookmarkedPosts, loading, loadMoreProps} = useMulti({
+    collectionName: "Posts",
+    terms: {
+      view: "myBookmarkedPosts",
+      limit: limit,
+    },
+    itemsPerPage: 20,
+    fragmentName: "PostsList",
+    skip: !currentUser?._id,
+  });
+  
+  // HACK: The results have limit/pagination which correctly reflects the order
+  // of currentUser.bookmarkedPostsMetadata, but within the limited result set
+  // the posts themselves may be out of order. Sort them. See also comments in
+  // the myBookmarkedPosts view.
+  let sortedBookmarkedPosts = sortBy(bookmarkedPosts,
+    post => -findIndex(
```

Comment:
Feels a bit spooky to rely on an iterator with O(n) runtime for sorting.  Assuming this is fine because of the pagination for first load, but any chance this blows up if a user has a few hundred bookmarks and decides to go digging?

Reply:
I think the constant factors are such that it would take a lot of clicks of the Load More button before it became an issue (and that's already O(n^2) in a sense, since each click of Load More reloads the entire expanded list).